Transfer ownership of files over ssh with sudo/su

benridley opened the pull request:
This PR addresses #614 by using the SSH user to assign a read-only ACL to the uploaded file, before switching to the sudo/su user and copying the file to the intended destination rather than attempting to `mv` it.

This allows a file uploaded by an unprivileged user to be 'handed over' to another unprivileged user. 

I've created this PR as a draft because there's a few issues worth discussing:
- This approach relies on the ability to use filesystem ACLs. Most modern filesystems support this but not all of them do, so I wanted to make sure this fits with pyinfra's philosophy. We can always try this and fall back to another approach as well.
- This introduces `cp` instead of `mv` when transferring the file out of the temp directory on upload, so for large files this could be a significant amount of extra storage consumed during the upload process.

I've also tried to update the tests, one of them updated ok but another is failing because it's asserting against the wrong call (it's using the `setfacl` call rather than the `cp` call). Happy to fix this up if I can get some assistance.

Commit:
Transfer ownership of files over ssh with sudo/su.

Adds a step to SSH file uploads that assigns read permissions via
ACLs to the sudo/su user, and uses cp instead of mv to allow the
the intended user to take ownership.

benridley commented:
Ah, apologies I meant to create this as a draft PR and now don't seem to be able to change it

Commit:
Use format to maintain python2 compatibility

Fizzadar reviewed (commented):
@benridley thank you for looking into this! I like the approach here, seems like a sensible solution to the problem. My only concern is the change to `cp` from `mv`, I'm wondering if there's a way to support using `mv` with acl's here; will do some testing locally over the next few days on that...

benridley commented:
Cheers @Fizzadar!

I've been using this in my fork of pyinfra for a week or so now for standing up dev environments fairly heavily, and I haven't hit any issues yet. 

I've also done some more investigation into the `mv` vs `cp` semantics that might help you in your investigation: The `setfacl` command demonstrated here actually does allow the other user to `mv` the file under normal circumstances. The problem is actually that the `/tmp` directory on most distributions has the sticky bit set. The `mv` command uses the `rename` syscall under the hood, and `rename` fails on directories with sticky bits set unless the uids match (see EPERM [on this link](https://man7.org/linux/man-pages/man2/rename.2.html)). So, when using /tmp unfortunately I can't see any way around using `cp`. 

The good news is that when the files are not on the same filesystem, `rename` doesn't work (see EXDEV in the link above), so `mv` actually does a copy and delete under the hood. Usually `/tmp` is mounted on its own filesystem, so in cases where a user is moving a file somewhere other than `/tmp` (which I would think would be most cases), and `/tmp` is just used as pyinfra's staging area, this change actually doesn't result in any change in behaviour provided we delete the temporary file after the `cp` operation.

Fizzadar commented:
@benridley that's incredibly helpful, thank you! I did recall there is actually already [a note on the operation docs](https://github.com/Fizzadar/pyinfra/blob/current/pyinfra/operations/files.py#L670-L672) that it's not suitable for large files due to potential copying.

Given the semantics of `mv`/`cp` I think using `cp` is fine here, if we're on the same filesystem the sticky bit makes it problematic and if separate filesystems `mv` will copy the file anyway. Switching to `cp` only keeps things nice and consistent and makes it possible to solve this problem.

Fizzadar commented on pyinfra/api/connectors/ssh.py:
```diff
@@ -420,8 +420,25 @@ def put_file(
 
+        # Make sure our sudo/su user can access the file
+        if su_user:
+            command = StringCommand('setfacl -m u:{0}:r'.format(su_user), temp_file)
```
```suggestion
            command = StringCommand('setfacl', '-m', 'u:{0}:r'.format(su_user), temp_file)
```

Fizzadar commented on pyinfra/api/connectors/ssh.py:
```diff
@@ -420,8 +420,25 @@ def put_file(
+        if su_user:
+            command = StringCommand('setfacl -m u:{0}:r'.format(su_user), temp_file)
+        elif sudo_user:
+            command = StringCommand('setfacl -m u:{0}:r'.format(sudo_user), temp_file)
```
```suggestion
            command = StringCommand('setfacl', '-m', 'u:{0}:r'.format(sudo_user), temp_file)
```

Minor nitpick on the style, makes no difference in output terms but consistent with `StringCommand` use!

Fizzadar commented on pyinfra/api/connectors/ssh.py:
```diff
@@ -420,8 +420,25 @@ def put_file(
+
         # Execute run_shell_command w/sudo and/or su_user
-        command = StringCommand('mv', temp_file, QuoteString(remote_filename))
+        command = StringCommand('cp', temp_file, QuoteString(remote_filename))
```
```suggestion
        command = StringCommand('cp', temp_file, QuoteString(remote_filename), '&&', 'rm', '-f', temp_file)
```

I think it makes sense to remove the temporary file here once the copy is complete.

benridley replied:
I think for this one we will need to do the removal outside the sudo session, otherwise we'll end up with the same problem as `mv` where we can't delete it due to the sticky bit. I'll give something a try tomorrow and make sure it works!

benridley replied:
I've added an `rm` outside the `sudo` session to remove the file now.

Fizzadar commented on tests/test_connectors/test_ssh.py:
```diff
@@ -540,7 +540,7 @@ def test_put_file_su_user_fail(self, fake_sftp_client, fake_ssh_client):
 
         fake_ssh_client().exec_command.assert_called_with((
-            "su centos -c 'sh -c '\"'\"'mv "
+            "su centos -c 'sh -c '\"'\"'cp "
```
This command will now be the `acl` one I believe - as the mock above is failing the first shell command executed. Changing this to assert for the acl command should cover the test.

Would also be good to have an addition test for the `cp` command failing vs. first ACL one, but I'm happy to add that in as well.

Fizzadar reviewed (commented):
This is awesome @benridley! Left a couple of minor things and the possible test fix :)

benridley committed:
Modify StringCommand for consistent styling.

Co-authored-by: Nick Barrett <nick@fizzadar.com>

benridley committed:
Modify StringCommand for consistent styling.

Co-authored-by: Nick Barrett <nick@fizzadar.com>

Commit:
Only run setfacl command if a sudo/su user is specified

Commit:
Merge branch 'feature_facl_ssh' of github.com:benridley/pyinfra into feature_facl_ssh

Commit:
Remove redundant chown command

The chown command is no longer required because we now cp the file as
the sudo/su user, so the file is created with the correct ownership.

Commit:
Remove temporary files after upload

Commit:
Modify test cases for new SSH upload commands

benridley commented:
Hi @Fizzadar, I've just pushed some new changes. Now the file is removed after the `cp` operation by running outside `sudo`. I tested this on the environment I've been working on and so far it's working as inteded.

I also fixed up the test cases so they are all passing, but you're correct that now the assertions are against the `setfacl` command, and now I've added the `rm` one of the other tests is asserting against that. 

If it's not too much pain would you mind assisting with adding the additional assertions? I'm happy to do so, but it might take me some time to get my head around it since I've not worked with pytest and mock etc before.

Take a look and let me know what you think, cheers!

Fizzadar committed:
Linting whitespaces.

Fizzadar committed:
Expand SSH connector tests to cover copy & acl failures.

Fizzadar commented:
@benridley thank you for updating this; I've just pushed updated tests for failures at each stage in https://github.com/Fizzadar/pyinfra/pull/616/commits/1a2f68640e073bbdd74df795e2d06b486222a04b.

benridley commented:
Awesome! Thanks @Fizzadar. Anything further you'd like to see on this PR?

Fizzadar reviewed (approved):
All good, thank you very much for working on this @benridley!
